vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go

```diff
@@ -192,13 +202,60 @@ type clusterStateFeeder struct {
 	memorySaveMode      bool
 	controllerFetcher   controllerfetcher.ControllerFetcher
 	recommenderName     string
+	historySource       HistorySource
+	promHistoryConfig   history.PrometheusHistoryProviderConfig
+}
+
+func (feeder *clusterStateFeeder) Init() error {
+	switch feeder.historySource {
+	case Checkpoints:
+		klog.Infof("Using checkpoints as a history provider")
+		feeder.initFromCheckpoints()
+	case Prometheus:
+		klog.Infof("Using prometheus as a history provider")
+
+		provider, promInitErr := history.NewPrometheusHistoryProvider(feeder.promHistoryConfig)
+		if promInitErr != nil {
+			klog.Errorf("Could not initialize history provider")
+			return promInitErr
+		}
+
+		historyInitErr := feeder.initFromHistoryProvider(provider)
+		if historyInitErr != nil {
+			klog.Errorf("Failed to load prometheus history")
+			return historyInitErr
+		}
+	case None:
+		klog.Infof("Running without a history provider")
+	default:
+		klog.Errorf("Wrong history provider option")
+		return fmt.Errorf("history provider option is not set. Supported values: prometheus, none, checkpoint")
+	}
+	return nil
 }
 
-func (feeder *clusterStateFeeder) InitFromHistoryProvider(historyProvider history.HistoryProvider) {
+func GetHistorySourceFromArg(historySource string) (HistorySource, error) {
+	switch historySource {
+	case "checkpoint":
+		return Checkpoints, nil
+	case "prometheus":
+		return Prometheus, nil
+	case "none":
+		return None, nil
+	default:
+		return Undefined, fmt.Errorf("storage option '%s' is not supported. Supported values: prometheus, none, checkpoint", historySource)
+	}
+}
+
+func (feeder *clusterStateFeeder) initFromHistoryProvider(historyProvider history.HistoryProvider) (historyInitError error) {
+	historyInitError = nil
 	klog.V(3).Info("Initializing VPA from history provider")
 	clusterHistory, err := historyProvider.GetClusterHistory()
 	if err != nil {
-		klog.Errorf("Cannot get cluster history: %v", err)
+		historyInitError = err
```

Comment:
Can you clarify why returning the error here right away doesn't work. I'm not sure if the historyInitError variable is really needed.
I think when a function returns a value and an error, as soon as the error is non-nil the caller should usually make no assumptions about the value. If you get an error this function should probably handle that error and return.

Reply:
Very valid point, I was thinking that more error flavours are possible with generalised `historyProvider`, but that is clearly overthinking, I'll push requested change when will have a free moment